File: src/main/java/no/novari/flyt/archive/gateway/dispatch/mapping/PartMappingService.java
package no.novari.flyt.archive.gateway.dispatch.mapping;

import no.fint.model.resource.Link;
import no.fint.model.resource.arkiv.noark.PartResource;
import no.novari.flyt.archive.gateway.dispatch.model.instance.PartDto;
import org.springframework.stereotype.Service;

@Service
public class PartMappingService {

    private final KontaktinformasjonMappingService kontaktinformasjonMappingService;
    private final AdresseMappingService adresseMappingService;

    public PartMappingService(
            KontaktinformasjonMappingService kontaktinformasjonMappingService,
            AdresseMappingService adresseMappingService
    ) {
        this.kontaktinformasjonMappingService = kontaktinformasjonMappingService;
        this.adresseMappingService = adresseMappingService;
    }

    public PartResource toPartResource(PartDto partDto) {
        if (partDto == null) {
            return null;
        }
        PartResource partResource = new PartResource();
        partDto.getPartNavn().ifPresent(partResource::setPartNavn);
        partDto.getPartRolle().map(Link::with).ifPresent(partResource::addPartRolle);
        partDto.getKontaktperson().ifPresent(partResource::setKontaktperson);
        partDto.getOrganisasjonsnummer().ifPresent(partResource::setOrganisasjonsnummer);
        partDto.getFodselsnummer().ifPresent(partResource::setFodselsnummer);
        partDto.getKontaktperson().ifPresent(partResource::setKontaktperson);
        partDto.getAdresse()
                .map(adresseMappingService::toAdresseResource)
                .ifPresent(partResource::setAdresse);
        partDto.getKontaktinformasjon()
                .map(kontaktinformasjonMappingService::toKontaktinformasjon)
                .ifPresent(partResource::setKontaktinformasjon);
        return partResource;
    }

}

File: src/main/kotlin/no/novari/flyt/archive/gateway/dispatch/mapping/PartMappingService.kt
package no.novari.flyt.archive.gateway.dispatch.mapping

import no.novari.fint.model.resource.Link
import no.novari.fint.model.resource.arkiv.noark.PartResource
import no.novari.flyt.archive.gateway.dispatch.model.instance.PartDto
import org.springframework.stereotype.Service

@Service
class PartMappingService(
    private val kontaktinformasjonMappingService: KontaktinformasjonMappingService,
    private val adresseMappingService: AdresseMappingService,
) {
    fun toPartResource(partDto: PartDto?): PartResource? =
        partDto?.let {
            PartResource().apply {
                partDto.partNavn?.let(::setPartNavn)
                partDto.partRolle?.let(Link::with)?.let(::addPartRolle)
                partDto.kontaktperson?.let(::setKontaktperson)
                partDto.organisasjonsnummer?.let(::setOrganisasjonsnummer)
                partDto.fodselsnummer?.let(::setFodselsnummer)
                partDto.adresse?.let(adresseMappingService::toAdresseResource)?.let(::setAdresse)
                partDto.kontaktinformasjon
                    ?.let(kontaktinformasjonMappingService::toKontaktinformasjon)
                    ?.let(::setKontaktinformasjon)
            }
        }
}
